sw/device/lib/dif/dif_spi_device.h

```diff
@@ -0,0 +1,65 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef SW_DEVICE_LIB_DIF_DIF_SPI_DEVICE_H_
+#define SW_DEVICE_LIB_DIF_DIF_SPI_DEVICE_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/dif/register.h"
+
+/**
+ * Initializes the SPI device located at |base_addr|.
+ *
+ * This function *must* be called on a particular |reg32_t| before calling any
+ * other functions in this header with that |reg32_t|; moreover, it *may not*
+ * be called more than once with any |reg32_t|. All deviations are Undefined
+ * Behavior.
+ *
+ * @param base_addr the base address for the SPI device registers.
+ */
+void spi_device_init(reg32_t base_addr);
+
+/**
+ * Attempts to send |len| bytes from the buffer pointed to by |data| over the
+ * initialized SPI device at |base_addr|.
+ *
+ * |data| *must* point to a buffer of at least |len| bytes of initialized data;
+ * not doing so is Undefined Behavior.
+ *
+ * @param base_addr the base address for the SPI device registers.
+ * @param data a contiguous buffer to copy from.
+ * @param len the length of the buffer |data| points to.
+ * @return the number of bytes actually sent.
+ */
+size_t spi_device_send(reg32_t base_addr, const void *data, size_t len);
```

Comment:
I think the format of the functions should be as suggested in the [spec](https://docs.google.com/document/d/1rMDHLQXSDH8CaUg0cID5ARGxdbflV61uk4crKWjhxs0/edit#):
bool timer_init(timer_t *timer, uintptr_t base_address);

* return bool
* have state structure per DIF (spi_t *spi)

I think it is useful to have "state" structure, and there is no point passing the base_address to any other functions apart from init.

Reply:
I don't want to return `bool` for reasons I bring up in your change. When I reviewed the spec, I did  not take that signature to be any sort of mandate.

However, I've added config and state types.

Reply:
I was thinking exactly the same, I think it could be useful to establish useful error codes like "Busy", "error", "success" and have them somewhere in a common DIF header. What do you think?